syntax = "proto3";

package ai.verta.modeldb.versioning;
option go_package = "github.com/VertaAI/modeldb/protos/gen/go/protos/public/modeldb/versioning";

option java_multiple_files = true;

import "modeldb/versioning/Enums.proto";

message DatasetBlob {
    oneof content {
        S3DatasetBlob s3 = 1;     // Blob of the type S3 dataset
        PathDatasetBlob path = 2; // Blob of the type path dataset
    }
}

message PathDatasetComponentBlob {
    string path = 1; // Full path to the file
    uint64 size = 2;
    uint64 last_modified_at_source = 3;

    string sha256 = 4;
    string md5 = 5;
}

message PathDatasetBlob {
    repeated PathDatasetComponentBlob components = 1;
}

message S3DatasetComponentBlob {
    PathDatasetComponentBlob path = 1;
    string s3_version_id = 2; //if s3 bucket versioning is enabled, this captures the version id
}

message S3DatasetBlob {
    repeated S3DatasetComponentBlob components = 1;
}

message DatasetDiff {
    oneof content {
        S3DatasetDiff s3 = 1;
        PathDatasetDiff path = 2;
    }
}

message PathDatasetComponentDiff {
    DiffStatusEnum.DiffStatus status = 1;
    PathDatasetComponentBlob A = 2;
    PathDatasetComponentBlob B = 3;
    PathDatasetComponentBlob C = 4;
}

message PathDatasetDiff {
    repeated PathDatasetComponentDiff components = 1;
}

message S3DatasetComponentDiff {
    DiffStatusEnum.DiffStatus status = 3;
    S3DatasetComponentBlob A = 4;
    S3DatasetComponentBlob B = 5;
    S3DatasetComponentBlob C = 6;
}

message S3DatasetDiff {
    repeated S3DatasetComponentDiff components = 1;
}
